returnn/tf/layers/basic.py

```diff
@@ -1645,6 +1645,98 @@ def get_out_data_from_opts(cls, name, sources, start=None, window_start=None, wi
     return out
 
 
+class RandIntLayer(LayerBase):
+  """
+  Generates random numbers using ``tf.random.uniform``
+  """
+  layer_class = "rand_int"
+
+  # noinspection PyUnusedLocal
+  def __init__(self, shape, maxval, minval=0, dtype="int32", seed=None, **kwargs):
+    """
+    :param tuple[DimensionTag|int]|list[DimensionTag|int] shape: desired shape of output tensor
+    :param int maxval: upper bound (exclusive) on range of random values
+    :param int minval: lower bound (inclusive) on range of random values
+    :param str dtype: type of the output. For random ints, int32 and int64 make sense, but could also be floats
+    :param int|None seed: random seed
+    """
+    from returnn.tf.util.data import DimensionTag
+    super(RandIntLayer, self).__init__(**kwargs)
+
+    seed = seed if seed is not None else self.network.random.randint(2 ** 31)
+
+    shape_parsed = []
+    batch_dim_axes = []
+    dyn_axes_sizes = []
+    for ax, s in enumerate(shape):
+      if isinstance(s, int):
+        shape_parsed.append(s)
+      elif isinstance(s, DimensionTag):
+        if s.kind == DimensionTag.Types.Batch:
+          batch_dim_axes.append(ax)
+          shape_parsed.append(self.get_batch_dim())
+        elif isinstance(s.dimension, int):
+          shape_parsed.append(s.dimension)
+        elif s.dimension is None:
+          assert s.dyn_size is not None
+          shape_parsed.append(tf.reduce_max(s.dyn_size))
+          dyn_axes_sizes.append(tf.fill(tf.shape(s.dyn_size), shape_parsed[-1]))
+        else:
+          raise Exception("%s: invalid dim tag %s" % (self, s))
+      else:
+        raise TypeError("%s: invalid dim %s" % (self, type(s)))
+    shape_parsed = tuple(shape_parsed)
+
+    if batch_dim_axes:
+      assert len(batch_dim_axes) == 1
+      self.output.size_placeholder = {i: size_placeholder for i, size_placeholder in enumerate(dyn_axes_sizes)}
+    else:
+      self.output.size_placeholder = {}
```

Comment:
The `size_placeholder` actually should not be set here, but already in `get_out_data_from_opts` below.

Reply:
Is there a particular reason for that here? Most layers set the `size_placeholder` in the init.

Reply:
I did it now anyway.

Reply:
No. Actually almost all layers set it in `get_out_data_from_opts`. Only those layers who create a *new* dyn size, they do it in `__init__` (because we don't want to modify the TF graph in `get_out_data_from_opts`, i.e. add new ops in there). But in your case, you don't add anything new. You just reuse existing dyn sizes / existing `DimensionTag`s.

Setting them in `get_out_data_from_opts` is important, because currently they carry the `DimensionTag` information for the dynamic axes, and we need this information for various functions.

Reply:
Alright, thanks!